Miscellaneous scrapping fixes

gergondet-woven opened the pull request:
Hello drake-ros maintainers,

Thank you for making this project available. We have been using it to build our software for the past months and we have built up a number of patches that we hope can be useful to everyone.

This PR is the first I'm opening and it contains 4 small patches related to various edge cases in the workspace scrapping that we have encountered.

Below is a short summary of the individual patches, I am happy to submit separate PRs if you prefer.

P1: Remove elements that have a condition attribute on ROS1

This is used by a number of packages that are trying to support both versions at the same time.

This patch is actually part of https://github.com/RobotLocomotion/drake-ros/pull/336 we actually had our own patch that does pretty much the same thing (including packages that have the `$ROS_VERSION == 2` condition instead of excluding the `$ROS_VERSION == 1`). However, that PR incorporates more changes this particular fix is not available on main yet so I hope it's ok to bring it in with this PR.

P2: Fix an issue with some Python packages that have multiple entries in the egg `top_level.txt` (seen in `launch_testing for example)

P3: Fix an issue with CMake's JSON where libraries' paths are escaped under some conditions (for example if the workspace path has a `~` character)

P4: Include `*.inc` files in C++ headers of library, afaik, only the `domain_bridge` package uses this extension

Thanks again for making bazel_ros2_rules and I'm looking forward to your feedback.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378)
<!-- Reviewable:end -->

zachfang commented:
@IanTheEngineer I assigned you as the reviewer for now just to ensure it's on our radar. Feel free to delegate otherwise.

gergondet-woven commented:
Hello @IanTheEngineer 

Thanks for the feedback.

> We try to wrap lines at 80 characters whenever possible. This applies to the other files in the PR as well.

Sorry about that, I have pushed an extra commit to wrap all lines here under 80 columns.

IanTheEngineer reviewed (changes requested):
I took another pass looking through the changes this evening. This looks quite good. I have a couple small questions and nits below. Thank you for iterating on this.

> Reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-:-OKZUQXx9-SxXVcROVDv:b-6u3xwi)* status: all files reviewed, 4 unresolved discussions (waiting on @gergondet-woven)

___
*[`bazel_ros2_rules/ros2/resources/ros2bzl/scraping/ament_python.py` line 17 at r2](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-OKZNmwL4TcDFTutIRlE:-OKZNmwL4TcDFTutIRlF:b9vvtn2) ([raw file](https://github.com/RobotLocomotion/drake-ros/blob/03b326a3670a902c1d9804ec86b87604634f5ba6/bazel_ros2_rules/ros2/resources/ros2bzl/scraping/ament_python.py#L17)):*
> ```Python
>     top_level = dist.read_text('top_level.txt')
>     top_level = top_level.rstrip('\n')
>     return str(dist._path), "\n".join(
> ```

Nit: Consider storing this string in a variable and returning that for readability. Also, consider adding a comment as to what this is fixing.
___
*[`bazel_ros2_rules/ros2/resources/ros2bzl/templates.py` line 156 at r2](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-OKZSfB43GawmaBIgoz9:-OKZSfB43GawmaBIgozA:b-5dmzmm) ([raw file](https://github.com/RobotLocomotion/drake-ros/blob/03b326a3670a902c1d9804ec86b87604634f5ba6/bazel_ros2_rules/ros2/resources/ros2bzl/templates.py#L156)):*
> ```Python
>         sandbox(top_level_i)
>         for _, top_level in properties.python_packages
>         for top_level_i in top_level.split("\n")
> ```

Nit: This nested list comprehension is quite clever, but somewhat hard to understand at a glance. Consider breaking it up for readability.
___
*[`bazel_ros2_rules/ros2/resources/ros2bzl/scraping/metadata.py` line 15 at r2](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-OKZPzjs1Akq0MXUnz7T:-OKZPzjs1Akq0MXUnz7U:bbbgwr6) ([raw file](https://github.com/RobotLocomotion/drake-ros/blob/03b326a3670a902c1d9804ec86b87604634f5ba6/bazel_ros2_rules/ros2/resources/ros2bzl/scraping/metadata.py#L15)):*
> ```Python
>                     elements_to_remove.append((parent, child))
>                 else:
>                     child.attrib.pop("condition")
> ```

Nit: Consider conforming to the single quote string convention used by the rest of the file.
___
*[`bazel_ros2_rules/ros2/resources/ros2bzl/scraping/metadata.py` line 15 at r2](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-OKZQiy18j4re9TfPt93:-OKZQiy18j4re9TfPt94:bgttsd7) ([raw file](https://github.com/RobotLocomotion/drake-ros/blob/03b326a3670a902c1d9804ec86b87604634f5ba6/bazel_ros2_rules/ros2/resources/ros2bzl/scraping/metadata.py#L15)):*
> ```Python
>                     elements_to_remove.append((parent, child))
>                 else:
>                     child.attrib.pop("condition")
> ```

Looking at the [ElementTree iter documentation](https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.Element.iter), it states that the behavior is undefined when iterating over a tree if it is then modified.

Would `child.attrib.pop('condition')` be considered a modification to the tree structure?
___
*[`bazel_ros2_rules/ros2/resources/cmake_tools/file_api.py` line 76 at r2](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-OKZOd-vAnNdghEMEnaR:-OKZOd-vAnNdghEMEnaS:b-t87vt6) ([raw file](https://github.com/RobotLocomotion/drake-ros/blob/03b326a3670a902c1d9804ec86b87604634f5ba6/bazel_ros2_rules/ros2/resources/cmake_tools/file_api.py#L76)):*
> ```Python
>                     # This escaping creates problem down the line,
>                     # "/usr/lib/libfoo.so" is *not* a path.
>                     # We remove the escaping here so they can be treated normally
> ```

This comment block is fantastic for conveying the issue and fix. Thanks for this!


<!-- Sent from Reviewable.io -->

gergondet-woven reviewed (commented):
Thanks again for the feedback and sorry for the delay but I finally got around to fixing the issues you raised!

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-:-OLrJcHAAGF-6IBkOcxy:bokwyd5)* status: 2 of 5 files reviewed, 2 unresolved discussions (waiting on @IanTheEngineer)

___
*[`bazel_ros2_rules/ros2/resources/ros2bzl/templates.py` line 156 at r2](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-OKZSfB43GawmaBIgoz9:-OLrHfjqC99DHbQJ4sgK:b-guo482) ([raw file](https://github.com/RobotLocomotion/drake-ros/blob/03b326a3670a902c1d9804ec86b87604634f5ba6/bazel_ros2_rules/ros2/resources/ros2bzl/templates.py#L156)):*
<details><summary><i>Previously, IanTheEngineer (Ian McMahon) wrote…</i></summary><blockquote>

Nit: This nested list comprehension is quite clever, but somewhat hard to understand at a glance. Consider breaking it up for readability.
</blockquote></details>

Done, hopefully it's clearer
___
*[`bazel_ros2_rules/ros2/resources/ros2bzl/scraping/ament_python.py` line 17 at r2](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-OKZNmwL4TcDFTutIRlE:-OLrHipg1sg9UFC3jUQA:b-8ttbng) ([raw file](https://github.com/RobotLocomotion/drake-ros/blob/03b326a3670a902c1d9804ec86b87604634f5ba6/bazel_ros2_rules/ros2/resources/ros2bzl/scraping/ament_python.py#L17)):*
<details><summary><i>Previously, IanTheEngineer (Ian McMahon) wrote…</i></summary><blockquote>

Nit: Consider storing this string in a variable and returning that for readability. Also, consider adding a comment as to what this is fixing.
</blockquote></details>

Done, I have split this into two assignments: one to retrieve the entries inside the the top_level file and one to resolve the path. I have also added a comment clarifying why we need to do this here
___
*[`bazel_ros2_rules/ros2/resources/ros2bzl/scraping/metadata.py` line 15 at r2](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-OKZPzjs1Akq0MXUnz7T:-OLrI0zb49n6AB0gc3NZ:bemoupz) ([raw file](https://github.com/RobotLocomotion/drake-ros/blob/03b326a3670a902c1d9804ec86b87604634f5ba6/bazel_ros2_rules/ros2/resources/ros2bzl/scraping/metadata.py#L15)):*
<details><summary><i>Previously, IanTheEngineer (Ian McMahon) wrote…</i></summary><blockquote>

Nit: Consider conforming to the single quote string convention used by the rest of the file.
</blockquote></details>

Sorry about that, I have changes all the double quotes in the patch with single quotes
___
*[`bazel_ros2_rules/ros2/resources/ros2bzl/scraping/metadata.py` line 15 at r2](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-OKZQiy18j4re9TfPt93:-OLrIA_kDMdtqwBjHXrp:bmuwv10) ([raw file](https://github.com/RobotLocomotion/drake-ros/blob/03b326a3670a902c1d9804ec86b87604634f5ba6/bazel_ros2_rules/ros2/resources/ros2bzl/scraping/metadata.py#L15)):*
<details><summary><i>Previously, IanTheEngineer (Ian McMahon) wrote…</i></summary><blockquote>

Looking at the [ElementTree iter documentation](https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.Element.iter), it states that the behavior is undefined when iterating over a tree if it is then modified.

Would `child.attrib.pop('condition')` be considered a modification to the tree structure?
</blockquote></details>

I don't think so, attributes are attached to the node so removing them should not modify the tree structure.


<!-- Sent from Reviewable.io -->

IanTheEngineer reviewed (approved):
Thank you for the iterations here! Things look great. I'll give this a test run tomorrow and we should be clear to merge.

> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-:-OO__08GE0KLpCNaELe5:b1ovyrd)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @gergondet-woven)



<!-- Sent from Reviewable.io -->

gergondet-woven commented:
Thanks for the heads up, I have rebased this branch on the current main to account for the changes.

The patches are not Humble specific and still apply to some packages in Jazzy.

> P2: Fix an issue with some Python packages that have multiple entries in the egg top_level.txt (seen in `launch_testing for example)

I have revised my patch for this since it was affected by #383. In Jazzy, the only package doing that - that I have installed locally at least - is `launch_testing`. Since #383 added an assert that `top_level.txt` would contain exactly 1 entry, this would fail with this package so I have changed the assert to allow 1 or more package and print a warning when it's more than 1.

ggould-tri reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378)* status: 2 of 6 files reviewed, 1 unresolved discussion (waiting on @IanTheEngineer)

___
*[`bazel_ros2_rules/ros2/resources/ros2bzl/scraping/python.py` line 26 at r4](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-OP64IkjB47BUau1gO1F:-OP64IkjB47BUau1gO1G:b-pit5ws) ([raw file](https://github.com/RobotLocomotion/drake-ros/blob/a9113f3f1756c8a148a78f3a620c0097808752ef/bazel_ros2_rules/ros2/resources/ros2bzl/scraping/python.py#L26)):*
> ```Python
>     if len(packages) > 1:
>         print(f"Multiple top level entries where found in {name}. "
>                "Only the first one will be considered")
> ```

BTW:  If the only place this is called is line 53 of this file, you could just return the whole list of packages and not build a singleton list around it there?
(I doubt this is especially important, but it would also probably net simplify the code)


<!-- Sent from Reviewable.io -->

ggould-tri reviewed (commented):
What's the current status on this PR?  I see one open discussion and some unclear CI status; I kicked the CI to unwedge that.  Who has the next action on this PR?

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378#-:-OPkVhbBE7tekkOuXw2q:biwdvcj)* status: 2 of 6 files reviewed, 1 unresolved discussion (waiting on @gergondet-woven and @IanTheEngineer)



<!-- Sent from Reviewable.io -->

gergondet-woven commented:
> Is this failure related?

Yes! This one was on me, I had a silly typo in that latest patch!

I pushed a fixed version where I verified locally that all tests passed :tada: 

Would you mind re-running the CI on this latest revision?

ggould-tri reviewed (approved):
> Reviewed 2 of 5 files at r2, 3 of 4 files at r4, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake-ros/378)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @gergondet-woven)



<!-- Sent from Reviewable.io -->
